server/src/citypay/citypay.service.ts

```diff
@@ -0,0 +1,112 @@
+import { Injectable } from '@nestjs/common';
+import * as puppeteer from 'puppeteer';
+import * as superagent from 'superagent';
+import { ConfigService } from '../config/config.service';
+
+const MAX_RETRIES = 5;
+const MS_APPLICANT_PORTAL = {
+  COOKIE_NAME: '.AspNet.ApplicationCookie',
+  SIGN_IN_PAGE: 'https://nycdcppfs.dynamics365portals.us/SignIn',
+  TOKEN_API: 'https://nycdcppfs.dynamics365portals.us/_services/auth/token',
+  CART_KEY_API: 'https://appservicedcpzapprod.azurewebsites.net/api/payment/getcartkey',
+
+   // security measure — they sign forms with this token
+  MICROSOFT_SIGN_IN_REQUEST_TOKEN: 'https://nycdcppfs.dynamics365portals.us/_layout/tokenhtml'
+}
+
+const CITY_PAY_LINK = 'https://a836-citypay.nyc.gov/citypay/retail/dcp-zap/processPreparedRetail';
+
+@Injectable()
+export class CitypayService {
+  constructor(
+    private readonly config: ConfigService,
+  ) {}
+
+  async generateCityPayLink(packageId) {
+    console.log('Generating new key...');
+    const newCartKey = await this.generateCartKey(packageId);
+
+    return this.createCartLink(newCartKey);
+  }
+
+  private createCartLink(CartKey) {
+    return `${CITY_PAY_LINK}?cartKey=${CartKey}`;
+  }
+
+  private async generateCartKey(packageId, retries = MAX_RETRIES) {
+    console.log('getting cart key...');
+
+    try {
+      const cookies = await this.stealCookies();
+      const microsoftPortalCookie = cookies.find(cookie => cookie.name === MS_APPLICANT_PORTAL.COOKIE_NAME)
+
+      // fetch the JWT to sign the next req
+      const MSPortalJWTRequest = await superagent
+        .get(MS_APPLICANT_PORTAL.TOKEN_API)
+        .query({ client_id: '29f7ab7e-e9bb-4b20-9b8f-4760a8313b28' })
+        .withCredentials()
+        .set('Cookie', `${MS_APPLICANT_PORTAL.COOKIE_NAME}=${microsoftPortalCookie.value}`)
+        .set('Accept-Encoding', 'gzip, deflate, br');
+      const cartKeyJWT = MSPortalJWTRequest.body.toString();
+
+      // fetch the CartKey used to identify the transaction on CityPay
+      const { body: { CartKey } } = await superagent
+        .post(MS_APPLICANT_PORTAL.CART_KEY_API)
+        .set('Authorization', `Bearer ${cartKeyJWT}`)
+        .send(`id=${packageId}`);
+
+      return CartKey;
+    } catch (e) {
+      console.log(`There was a problem: ${e.toString()}`);
+      console.log('Trying again...');
+
+      if (retries) {
+        return this.generateCartKey(packageId, retries - 1);
+      } else {
+        throw new Error(`Could not generate the key for some reason: ${e.toString}`);
+      }
+    }
+  }
+
+  // generates cookies by headless login to MS Applicant Portal
+  private async stealCookies(retries = MAX_RETRIES) {
```

Comment:
stealCookies isn't the clearest name... Maybe something like "getMsPortalCookies"